File: windows/src/core/timecode.js
// Timecode math shared by the LTC decoder and the MTC generator.
// JavaScript port of Sources/Core/Timecode.swift.

export const FrameRate = Object.freeze({
  fps24: { id: 0, fps: 24, drop: false, period: 1 / 24, label: '24 fps', short: '24' },
  fps25: { id: 1, fps: 25, drop: false, period: 1 / 25, label: '25 fps', short: '25' },
  fps2997DF: { id: 2, fps: 30, drop: true, period: 1001 / 30000, label: '29.97 DF', short: '29.97 DF' },
  fps30: { id: 3, fps: 30, drop: false, period: 1 / 30, label: '30 fps', short: '30' },
});

export const allRates = [FrameRate.fps24, FrameRate.fps25, FrameRate.fps2997DF, FrameRate.fps30];
export const rateById = (id) => allRates.find((r) => r.id === id);

/** Frames in 24 hours. */
export const framesPerDay = (rate) => (rate.drop ? 2589408 : 24 * 3600 * rate.fps);

export function tc(hours, minutes, seconds, frames) {
  return { hours, minutes, seconds, frames };
}

export const tcEqual = (a, b) =>
  a.hours === b.hours && a.minutes === b.minutes && a.seconds === b.seconds && a.frames === b.frames;

/** Timecode → absolute frame count since 00:00:00:00. */
export function frameCount(t, rate) {
  let count = ((t.hours * 60 + t.minutes) * 60 + t.seconds) * rate.fps + t.frames;
  if (rate.drop) {
    const totalMinutes = t.hours * 60 + t.minutes;
    count -= 2 * (totalMinutes - Math.floor(totalMinutes / 10));
  }
  return count;
}

/** Absolute frame count → timecode, wrapping at 24 h. */
export function fromFrameCount(count, rate) {
  const perDay = framesPerDay(rate);
  let n = ((count % perDay) + perDay) % perDay;
  if (rate.drop) {
    const d = Math.floor(n / 17982);
    const m = n % 17982;
    n += 18 * d + (m < 2 ? 0 : 2 * Math.floor((m - 2) / 1798));
  }
  const fps = rate.fps;
  return tc(Math.floor(n / (fps * 3600)) % 24, Math.floor(n / (fps * 60)) % 60, Math.floor(n / fps) % 60, n % fps);
}

const two = (v) => (v < 10 ? '0' + v : '' + v);
export function format(t, dropFrame = false) {
  return `${two(t.hours)}:${two(t.minutes)}:${two(t.seconds)}${dropFrame ? ';' : ':'}${two(t.frames)}`;
}

function parts(text) {
  const p = text.trim().split(/[:;.]/).map((s) => s.trim());
  if (p.length !== 4 || p.some((s) => !/^\d+$/.test(s))) return null;
  return p.map(Number);
}

/** Parses a signed "HH:MM:SS:FF" duration into a frame count (no drop-frame adjustment). */
export function parseOffset(text, rate) {
  let s = text.trim();
  let sign = 1;
  if (s.startsWith('-')) { sign = -1; s = s.slice(1); } else if (s.startsWith('+')) { s = s.slice(1); }
  const p = parts(s);
  if (!p || p[1] >= 60 || p[2] >= 60 || p[3] >= rate.fps) return null;
  return sign * (((p[0] * 60 + p[1]) * 60 + p[2]) * rate.fps + p[3]);
}

/** Parses "HH:MM:SS:FF" as a position; non-existent drop-frame labels move to frame 2. */
export function parseTimecode(text, rate) {
  const p = parts(text);
  if (!p || p[0] >= 24 || p[1] >= 60 || p[2] >= 60 || p[3] >= rate.fps) return null;
  const t = tc(p[0], p[1], p[2], p[3]);
  if (rate.drop && t.seconds === 0 && t.minutes % 10 !== 0 && t.frames < 2) t.frames = 2;
  return t;
}

File: Sources/Core/Timecode.swift
// Timecode math shared by the LTC decoder and the MTC generator.
// Platform-independent: no Apple frameworks used here.

enum FrameRate: Int, CaseIterable {
    case fps24 = 0
    case fps25 = 1
    case fps2997DF = 2
    case fps30 = 3

    /// Rate code used in MTC quarter-frame piece 7 and the full-frame message.
    var mtcCode: UInt8 { UInt8(rawValue) }

    /// Frames per timecode second (the frame-number wrap point).
    var framesPerSecond: Int {
        switch self {
        case .fps24: return 24
        case .fps25: return 25
        case .fps2997DF, .fps30: return 30
        }
    }

    var isDropFrame: Bool { self == .fps2997DF }

    /// Nominal real-time frame period in seconds.
    var nominalPeriod: Double {
        switch self {
        case .fps24: return 1.0 / 24.0
        case .fps25: return 1.0 / 25.0
        case .fps2997DF: return 1001.0 / 30000.0
        case .fps30: return 1.0 / 30.0
        }
    }

    /// Number of frames in 24 hours.
    var framesPerDay: Int {
        isDropFrame ? 2_589_408 : 24 * 3600 * framesPerSecond
    }

    var label: String {
        switch self {
        case .fps24: return "24 fps"
        case .fps25: return "25 fps"
        case .fps2997DF: return "29.97 DF"
        case .fps30: return "30 fps"
        }
    }
}

struct Timecode: Equatable {
    var hours: Int
    var minutes: Int
    var seconds: Int
    var frames: Int

    init(hours: Int, minutes: Int, seconds: Int, frames: Int) {
        self.hours = hours
        self.minutes = minutes
        self.seconds = seconds
        self.frames = frames
    }

    /// Converts timecode to an absolute frame count since 00:00:00:00.
    func frameCount(rate: FrameRate) -> Int {
        let fps = rate.framesPerSecond
        var count = ((hours * 60 + minutes) * 60 + seconds) * fps + frames
        if rate.isDropFrame {
            let totalMinutes = hours * 60 + minutes
            count -= 2 * (totalMinutes - totalMinutes / 10)
        }
        return count
    }

    /// Converts an absolute frame count back to timecode, wrapping at 24h.
    init(frameCount: Int, rate: FrameRate) {
        let perDay = rate.framesPerDay
        var n = ((frameCount % perDay) + perDay) % perDay
        if rate.isDropFrame {
            let d = n / 17982
            let m = n % 17982
            n += 18 * d + (m < 2 ? 0 : 2 * ((m - 2) / 1798))
        }
        let fps = rate.framesPerSecond
        frames = n % fps
        seconds = (n / fps) % 60
        minutes = (n / (fps * 60)) % 60
        hours = (n / (fps * 3600)) % 24
    }

    /// Parses "HH:MM:SS:FF" (also accepts ';' or '.' separators and a leading '-').
    /// Returns a signed frame count, or nil if malformed.
    static func parseOffset(_ text: String, rate: FrameRate) -> Int? {
        var s = text.trimmingSpaces()
        var sign = 1
        if s.hasPrefix("-") { sign = -1; s.removeFirst() } else if s.hasPrefix("+") { s.removeFirst() }
        let parts = s.split(whereSeparator: { $0 == ":" || $0 == ";" || $0 == "." }).map { Int($0) }
        guard parts.count == 4, parts.allSatisfy({ $0 != nil }) else { return nil }
        let p = parts.map { $0! }
        guard p[1] < 60, p[2] < 60, p[3] < rate.framesPerSecond else { return nil }
        // Offsets are plain durations, so no drop-frame adjustment.
        let fps = rate.framesPerSecond
        return sign * (((p[0] * 60 + p[1]) * 60 + p[2]) * fps + p[3])
    }

    /// Parses "HH:MM:SS:FF" as a timecode position. Drop-frame labels that don't exist
    /// (frames 0 and 1 of most minutes) are moved to frame 2.
    static func parse(_ text: String, rate: FrameRate) -> Timecode? {
        let parts = text.split(whereSeparator: { $0 == ":" || $0 == ";" || $0 == "." }).map { Int(String($0).trimmingSpaces()) }
        guard parts.count == 4, parts.allSatisfy({ $0 != nil }) else { return nil }
        let p = parts.map { $0! }
        guard p[0] < 24, p[1] < 60, p[2] < 60, p[3] < rate.framesPerSecond, p.allSatisfy({ $0 >= 0 }) else { return nil }
        var tc = Timecode(hours: p[0], minutes: p[1], seconds: p[2], frames: p[3])
        if rate.isDropFrame, tc.seconds == 0, tc.minutes % 10 != 0, tc.frames < 2 { tc.frames = 2 }
        return tc
    }

    func formatted(dropFrame: Bool = false) -> String {
        func two(_ v: Int) -> String { v < 10 ? "0\(v)" : "\(v)" }
        return "\(two(hours)):\(two(minutes)):\(two(seconds))\(dropFrame ? ";" : ":")\(two(frames))"
    }
}

private extension String {
    func trimmingSpaces() -> String {
        var s = Substring(self)
        while s.first == " " { s.removeFirst() }
        while s.last == " " { s.removeLast() }
        return String(s)
    }
}
